core/sdk-core/src/main/java/software/amazon/awssdk/core/internal/interceptor/RequestCompressionInterceptor.java

```diff
@@ -0,0 +1,292 @@
+import java.util.function.Supplier;
+import software.amazon.awssdk.annotations.SdkInternalApi;
+import software.amazon.awssdk.core.SdkSystemSetting;
+import software.amazon.awssdk.core.async.AsyncRequestBody;
+import software.amazon.awssdk.core.compression.CompressionType;
+import software.amazon.awssdk.core.compression.Compressor;
+import software.amazon.awssdk.core.exception.SdkClientException;
+import software.amazon.awssdk.core.interceptor.Context;
+import software.amazon.awssdk.core.interceptor.ExecutionAttributes;
+import software.amazon.awssdk.core.interceptor.ExecutionInterceptor;
+import software.amazon.awssdk.core.interceptor.SdkExecutionAttribute;
+import software.amazon.awssdk.core.interceptor.SdkInternalExecutionAttribute;
+import software.amazon.awssdk.core.sync.RequestBody;
+import software.amazon.awssdk.http.ContentStreamProvider;
+import software.amazon.awssdk.http.SdkHttpFullRequest;
+import software.amazon.awssdk.http.SdkHttpRequest;
+import software.amazon.awssdk.profiles.ProfileFile;
+import software.amazon.awssdk.profiles.ProfileFileSystemSetting;
+import software.amazon.awssdk.profiles.ProfileProperty;
+import software.amazon.awssdk.utils.IoUtils;
+
+/**
+ * Interceptor to handles compression of requests whose operations are marked with the "requestCompression" C2J trait.
+ * Compression of payload will be completed prior to checksum calculation. The Content-Encoding header will be updated
+ * accordingly.
+ */
+@SdkInternalApi
+public class RequestCompressionInterceptor implements ExecutionInterceptor {
+
+    private static final int DEFAULT_MIN_COMPRESSION_SIZE = 10_240;
+    private static final int MIN_COMPRESSION_SIZE_LIMIT = 10_485_760;
+    private static final Supplier<ProfileFile> PROFILE_FILE = ProfileFile::defaultProfileFile;
+    private static final String PROFILE_NAME = ProfileFileSystemSetting.AWS_PROFILE.getStringValueOrThrow();
+
+    @Override
+    public SdkHttpRequest modifyHttpRequest(Context.ModifyHttpRequest context, ExecutionAttributes executionAttributes) {
+        if (!shouldCompress(context, executionAttributes)) {
+            return context.httpRequest();
+        }
+
+        Compressor compressor = resolveCompressionType(executionAttributes);
+        if (executionAttributes.getAttribute(SdkInternalExecutionAttribute.REQUEST_COMPRESSION).isStreaming()) {
+            return updateContentEncodingHeader(context.httpRequest(), compressor);
+        }
+
+        SdkHttpFullRequest sdkHttpFullRequest = (SdkHttpFullRequest) context.httpRequest();
+        InputStream inputStream = sdkHttpFullRequest.contentStreamProvider().get().newStream();
+        InputStream compressedStream = compressor.compress(inputStream);
+        SdkHttpRequest sdkHttpRequest =
+            sdkHttpFullRequest.toBuilder()
+                              .contentStreamProvider(() -> compressedStream)
+                              .build();
+        sdkHttpRequest = updateContentEncodingHeader(sdkHttpRequest, compressor);
```

Comment:
This is is only doing a one-time compression of the wrapped stream. The new content stream provider should be wrapping the existing provider, something like this:

```
ContentStreamProvider compressedStreamProvider = () -> compressor.compress(wrappedProvider.newStream());
```